plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,270 @@
+#include <uk/assert.h>
+#include <kvm-arm/mm.h>
+#include <kvm-arm/smp.h>
+#include <kvm/intctrl.h>
+#include <arm/cpu.h>
+#include <uk/arch/limits.h>
+#include <stdbool.h>
+#include <uk/plat/bootstrap.h>
+#include <uk/sched.h>
+#include <uk/plat/time.h>
+
+#include <uk/plat/memory.h>
+#include <uk/plat/io.h>
+#include <uk/plat/lcpu.h>
+#include <arm/psci.h>
+#include <ofw/fdt.h>
+
+static __u64 smp_numcores;
+__u64 bspdone;
+__u64 smp_aps_started; // APs that are started and reached their entry function
+__u64 smp_aps_running; // APs that are started and aware of bspdone = 1
+
+struct arm64_cpu {
+    uint64_t id;
+	ukplat_lcpu_entry_t entry;
+    void *stackp;
+    struct uk_list_head fnlist;
+};
+const size_t arm64_cpu_size = sizeof(struct arm64_cpu);
+
+struct arm64_cpu cpus[CONFIG_MAX_CPUS];
+
+static __lcpuid bspid;
+
+static void ndelay(__u64 nsec)
+{
+       __nsec until = ukplat_monotonic_clock() + nsec;
+
+       while (until > ukplat_monotonic_clock())
+               ukplat_lcpu_halt_to(until);
+}
+
+static void mdelay(__u64 msec)
+{
+       ndelay(msec*1000*1000);
+}
+
+int smp_init(void)
+{
+	int fdt_cpu;
+	int naddr, cell;
+	const fdt32_t *prop;
+	uint64_t index;
+	int subnode;
+
+	/* Init the cpu_possible_map */
+	for (index = 0; index < CONFIG_MAX_CPUS; index++)
+        cpus[index].id = -1;
+
+	/* Search for assigned VM cpus in DTB */
+	fdt_cpu = fdt_path_offset(_libkvmplat_cfg.dtb, "/cpus");
+	if (fdt_cpu < 0)
+		uk_pr_warn("cpus node is not found in device tree\n");
+
+	/* Get address,size cell */
+	prop = fdt_getprop(_libkvmplat_cfg.dtb, fdt_cpu, "#address-cells", &cell);
+	if (prop != NULL)
+		naddr = fdt32_to_cpu(prop[0]);
+	if (naddr < 0 || naddr >= FDT_MAX_NCELLS) {
+		UK_CRASH("Could not find cpu address!\n");
+		return -1;
+	}
+
+	/* Search all the cpu nodes in DTB */
+	index = 0;
+	fdt_for_each_subnode(subnode, _libkvmplat_cfg.dtb, fdt_cpu) {
+		const struct fdt_property *prop;
+		int prop_len = 0;
+
+		prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"enable-method", NULL);
+        if (!prop) {
+            uk_pr_err("No method found!\n");
+            return -1;
+        }
+        else if (strcmp(prop->data, "psci")) {
+            uk_pr_err("Only support psci method!(%s)\n",
+				prop->data);
+			return -1;
+        }
+
+		prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"device_type", &prop_len);
+		if (!prop)
+			continue;
+		if (prop_len < 4)
+			continue;
+		if (strcmp(prop->data, "cpu"))
+			continue;
+
+		prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"reg", &prop_len);
+		if (prop == NULL || prop_len <= 0) {
+			uk_pr_err("Error when searching reg property\n");
+			return -1;
+		}
+
+		cpus[index].id = fdt_reg_read_number((const fdt32_t *)prop->data,
+						naddr);
+        uk_pr_info("Initialized core %d\n", cpus[index].id);
+		index++;
+	}
+	smp_numcores = index;
+
+    bspid = ukplat_lcpu_id();
+	uk_pr_info("Bootstrapping processor has the ID %d\n", bspid);
+
+    
```

Comment:
File indentation needs to be fixed. There are some extra tabs on blank lines and some mixing of TAB vs Space indentation....

Reply:
Yes, indentation is broken in various places. I already fixed them locally for me (cannot concentrate on the important things otherwise 😄). @razvanvirtan Before you start changing things: I will provide a patch, also with some other minor changes as basis for discussion. So keep the PR as-is until then.

Reply:
Sorry for this, I think some of the problems come because I've taken code from Justin's sources and haven't checked how it was indented (it looked good on my local).